db/python/tables/cohort.py

```diff
@@ -0,0 +1,195 @@
+    template_id: GenericFilter[int] | None = None
+    timestamp: GenericFilter[datetime.datetime] | None = None
+    project: GenericFilter[ProjectId] | None = None
+
+
+@dataclasses.dataclass(kw_only=True)
+class CohortTemplateFilter(GenericFilterModel):
+    """
+    Filters for CohortTemplate
+    """
+
+    id: GenericFilter[int] | None = None
+    name: GenericFilter[str] | None = None
+    description: GenericFilter[str] | None = None
+    criteria: GenericFilter[dict] | None = None
+    project: GenericFilter[ProjectId] | None = None
+
+
+class CohortTable(DbBase):
+    """
+    Capture Cohort table operations and queries
+    """
+
+    table_name = 'cohort'
+    common_get_keys = [
+        'id',
+        'name',
+        'template_id',
+        'description',
+        'author',
+        'project',
+    ]
+
+    template_keys = [
+        'id',
+        'name',
+        'description',
+        'criteria',
+        'project'
+    ]
+
+    async def query(self, filter_: CohortFilter):
+        """Query Cohorts"""
+        wheres, values = filter_.to_sql(field_overrides={})
+        if not wheres:
+            raise ValueError(f'Invalid filter: {filter_}')
+
+        common_get_keys_str = ','.join(self.common_get_keys)
+        _query = f"""
+        SELECT {common_get_keys_str}
+        FROM cohort
+        WHERE {wheres}
+        """
+
+        rows = await self.connection.fetch_all(_query, values)
+        cohorts = [Cohort.from_db(dict(row)) for row in rows]
+        return cohorts
+
+    async def get_cohort_sequencing_group_ids(self, cohort_id: int) -> list[int]:
+        """
+        Return all sequencing group IDs for the given cohort.
+        """
+
+        _query = """
+        SELECT sequencing_group_id FROM cohort_sequencing_group WHERE cohort_id = :cohort_id
+        """
+        rows = await self.connection.fetch_all(_query, {'cohort_id': cohort_id})
+        return [row['sequencing_group_id'] for row in rows]
+
+    async def query_cohort_templates(self, filter_: CohortTemplateFilter):  # TODO: Move this to its own class?
+        """Query CohortTemplates"""
+        wheres, values = filter_.to_sql(field_overrides={})
+        if not wheres:
+            raise ValueError(f'Invalid filter: {filter_}')
+
+        common_get_keys_str = ','.join(self.template_keys)
+        _query = f"""
+        SELECT {common_get_keys_str}
+        FROM cohort_template
+        WHERE {wheres}
+        """
+
+        rows = await self.connection.fetch_all(_query, values)
+        cohort_templates = [CohortTemplateModel.from_db(dict(row)) for row in rows]
+        return cohort_templates
+
+    async def get_cohort_template(self, template_id: int):
+        """
+        Get a cohort template by ID
+        """
+        _query = """
+        SELECT id as id, criteria as criteria FROM cohort_template WHERE id = :template_id
+        """
+        template = await self.connection.fetch_one(_query, {'template_id': template_id})
```

Comment:
If there are no rows with the requested `template_id`, this returns None and the `return` below explodes with `TypeError: 'NoneType' object is not subscriptable`. We should probably check this and at least return a better exception.

Reply:
Awesome catch, thanks @jmarshall! Have pushed a more graceful exit.

Reply:
Thanks, LGTM.

Reply:
Actually, looking at this again for more than a few seconds :smile:, I think here at the table level we should probably return `None` in this case rather than failing with an exception. It'll be dealt with and an exception raised by the only current caller in _layers/cohort.py_.

Or if we stick with an exception, it should be metamist's NotFoundError. And for bonus points, either way the docstring should say what happens in the not found case.

Reply:
Returned None instead of error :)

Reply:
Thanks!